src/cdomains/congruenceClosure.ml

```diff
@@ -0,0 +1,1275 @@
+    (* process all atoms in increasing order *)
+    let atoms =
+      List.sort (fun el1 el2 ->
+          let v1, z1 = TUF.find_no_pc uf el1 in
+          let v2, z2 = TUF.find_no_pc uf el2 in
+          let repr_compare = TUF.compare_repr (v1, z1) (v2, z2)
+          in
+          if repr_compare = 0 then T.compare el1 el2 else repr_compare) atoms in
+    let add_atom_to_map (min_representatives, queue, uf) a =
+      let rep, offs = TUF.find_no_pc uf a in
+      if not (mem rep min_representatives) then
+        (add rep (a, offs) min_representatives, queue @ [rep], uf)
+      else (min_representatives, queue, uf)
+    in
+    let (min_representatives, queue, uf) = List.fold_left add_atom_to_map (empty, [], uf) atoms
+    (* compute the minimal representative of all remaining edges *)
+    in update_min_repr (uf, set, map) min_representatives queue
+
+  let compute_minimal_representatives a = Timing.wrap "c2po-compute-min-repr" compute_minimal_representatives a
+
+  (** Computes the initial map of minimal representatives.
+        It maps each element `e` in the set to `(e, 0)`. *)
+  let initial_minimal_representatives set =
+    List.fold_left (fun map element -> add element (element, Z.zero) map) empty (SSet.elements set)
+end
+
+module Lazy =
+struct
+  include Lazy
+  let hash x y = 0
+end
+
+(**
+   This is the type for the abstract domain.
+
+   - `uf` is the union find tree
+
+   - `set` is the set of terms that are currently considered.
+     It is the set of terms that have a mapping in the `uf` tree.
+
+   - `map` maps reference variables v to a map that maps integers z to terms that are equivalent to *(v + z).
+     It represents the transitions of the quantitative finite automata.
+
+   - `normal_form` is the unique normal form of the domain element, it is a lazily computed when needed and stored such that it can be used again later.
+
+   - `diseq` represents the block disequalities. It is a map that maps each term to a set of terms that are definitely in a different block.
+
+   - `bldis` represents the disequalities. It is a map from a term t1 to a map from an integer z to a set of terms T, which represents the disequality t1 != z + t2 for each t2 in T.
+*)
+type t = {uf: TUF.t;
+          set: SSet.t;
+          map: LMap.t;
+          normal_form: T.v_prop list Lazy.t[@compare.ignore][@eq.ignore][@hash.ignore];
+          diseq: Disequalities.t;
+          bldis: BlDis.t}
+[@@deriving eq, ord, hash]
+
+let show_conj list = List.fold_left
+    (fun s d -> s ^ "\t" ^ T.show_prop d ^ ";\n") "" list
+
+(** Returns a list of all the transition that are present in the automata. *)
+let get_transitions (uf, map) =
+  List.concat_map (fun (t, zmap) ->
+      (List.map (fun (edge_z, res_t) ->
+           (edge_z, t, TUF.find_no_pc uf res_t)) @@
+       (LMap.zmap_bindings zmap)))
+    (LMap.bindings map)
+
+let exactly_equal cc1 cc2 =
+  cc1.uf == cc2.uf && cc1.map == cc2.map && cc1.diseq == cc2.diseq && cc1.bldis == cc2.bldis
```

Comment:
Why does one not check for structural equality here, as well? 
Or is it so that if these are not at the same address, one may as well check for semantic equality otherwise?